### ISSUE TITLE
Support Streaming Responses

### ISSUE THREAD

**Issue by shirshak55** [httptoolkit/httptoolkit#312]
Firebase etc. have streaming responses.

`https://firestore.googleapis.com/google.firestore.v1.Firestore/Listen/channel`

It would be nice if HttpToolKit can display it.

**Comment by pimterry** [comment]
Yes that would definitely be good! This relates to [gRPC](https://github.com/httptoolkit/httptoolkit/issues/44) support, but it would also be useful elsewhere.

The way this works currently is that Mockttp records all response data ([here](https://github.com/httptoolkit/mockttp/blob/a1e24040f645493d522c9a7f3798713d6e8888ea/src/util/request-utils.ts#L293-L404)) waits until the whole response has been sent back to the client (status code, headers & body) and then fires a single event with that data ([here](https://github.com/httptoolkit/mockttp/blob/main/src/server/mockttp-server.ts#L664)), but obviously that doesn't work very well if the response stays open and slowly streams data for a long time though (and with gRPC, I think it can even be streaming in both directions, to make this extra complicated).

This is solvable, but we need to decide how... I guess maybe we start an interval (maybe running every 1 second?) after we start sending the request & response bodies, and if we haven't finished yet, we fire an event with the latest partial body data? That could work, and should be easy enough to handle and show in the UI afterwards. Updating every second should make data appear fast enough for most use cases. Or we could do a cleverer debouncing kind of approach: fire partial data almost immediately, then fire throttled events, so that events for new data are usually sent immediately on relatively quiet connections (<1 new chunk of data per second), but very frequent updates (>1 new chunk per second) are batched up for efficiency.

Notably, for gRPC streaming calls it's possible that request & response bodies can be going both ways at the same time - this is [bidirectional streaming rpc](https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc). We should make sure it's possible to handle that case, which means firing request & response events with enough information that you can work out the overall order of data sent and received (so maybe every event is a series of chunks where each one has a timestamp, so we can work out the detailed timing given both request & response messages).

Compression needs some thought - in theory we can decompress a partially received body that's gzipped, but I've never tried it and there's probably some details to figure out there (or we just ignore it for now, and we never stream compressed data).

Anyway, that's just some notes for implementing this later. I'm not going to have time to do this immediately (currently busy adding interception for WebRTC) but if you or anybody else is especially interested then you're welcome to start work on it! The first step would be creating a `response-initiated` event in Mockttp, which fires when the status & headers have been sent, but before the body is sent, just like the existing `request-initiated` event. Then we'd need to extend this further to test some approaches to body streaming events, probably announcing the latest body data at frequent intervals as above. Just adding that first event would be a useful small change though. Once all the Mockttp parts are in to make the data available, the UI work should be easy.

**Comment by strikerlulu** [comment]
@pimterry any branch where i can try the WIP for this ?

**Comment by pimterry** [comment]
Hi @strikerlulu, there's no complete end-to-end version that's runnable on a branch yet. If you're interested in getting involved, the current state is:

* The UI currently listens to request-initiated, request & response events from the internal proxy ([here](https://github.com/httptoolkit/httptoolkit-ui/blob/df61ddb7bd107c984dd68c33cfb7a3cc25f5f321/src/model/events/events-store.ts#L60-L63)). These don't fire until the final state is complete, and they include the entire body.
* The internal proxy now supports `response-initiated` and `request/response-body-data` events, which fire separately: when the response starts (before the body is available) and then for each request/response chunk as it's received.
* The UI needs to:
  * Refactor the UI HTTP data model [here](https://github.com/httptoolkit/httptoolkit-ui/blob/df61ddb7bd107c984dd68c33cfb7a3cc25f5f321/src/model/http/http-exchange.ts#L107) because it's no longer just request started -> completed -> got response. There's more steps, and more cases where they can be independent in confusing ways (e.g. request body partially received but still streaming, but the response body is fully completed, or one side aborted part-way but the other not). The various methods like `hasResponseBody` (and how they're used to render the UI) need rethinking now that we have partial bodies etc.
  * Listen to the new events where available and feed that data into the updated data model, including incrementally building up the partial body data.
  * Update the UI to show the resulting states of the model correctly.
  * Extend the UI to provide some indicator when the body is partially available but not actually completed (still streaming, or just waiting e.g. for long polling cases).
  * To avoid breaking when using the new UI with an old server, listen to the old events as well if the new events aren't available, and somehow feed that into the new UI data model.

If you're interested in working on any of that, PRs are very welcome!

**Comment by vegerot** [comment]
👍 for this. One additional UX angle not fully covered above:

When a client closes an SSE (`text/event-stream`) socket at end-of-stream, the exchange is shown as a red **"This exchange was not completed successfully"** with `passthrough-error:ECONNRESET`. This is indistinguishable from a genuine upstream failure, even though the response streamed and was consumed successfully.

Evidence from an affected exchange — headers/body streamed for ~5.5s *before* the abort, while plain JSON calls to the same host over the same proxy return `200` normally:

```json
"tags": ["passthrough-error:ECONNRESET"],
"timing": {
  "startTimestamp": 331363.293,
  "headersSentTimestamp": 335712.120,
  "abortedTimestamp": 341176.325
}
```

Beyond showing partial streamed data (the core of this issue), it'd help to distinguish "socket RST *after* a response began streaming" from "connection failed before any response," and use a softer, non-error status for the former.
